Propose changes to resource ID section

aphillips opened the pull request:
This is I18N-ACTION-1208.

Add text.
Fix reference to published version of IRI.
Add reference to WHATWG URL.
Clean up one bit of mustard.


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/aphillips/bp-i18n-specdev/pull/80.html" title="Last updated on Oct 6, 2022, 3:16 PM UTC (ab922ad)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/bp-i18n-specdev/80/edde554...aphillips:ab922ad.html" title="Last updated on Oct 6, 2022, 3:16 PM UTC (ab922ad)">Diff</a>

aphillips committed:
Propose changes to resource ID section

This is I18N-ACTION-1208.

Add text.
Fix reference to published version of IRI.
Add reference to WHATWG URL.
Clean up one bit of mustard.

aphillips committed:
Merge branch 'w3c:gh-pages' into i18n-action-1208

aphillips committed:
Edits from discussion of 2022-09-29

Reformulated entire section. Imported text from charmod-resid where useful.

aphillips committed:
Merge branch 'w3c:gh-pages' into i18n-action-1208

klensin commented:
It may take me a bit of time to completely review the spec (reasons may have been clear from my reaction on the call today), but I immediately noticed one thing.  Part of the text reads" "... encoding non-ASCII characters as %HH escapes..." In reality, %HH escapes died (or became ambiguous) the moment non-ASCII characters on the web stopped being assumed to be coded in ISO 8859-1, where a character was still a single octet.    The description of how things are escaped should be folded together with the next sentence, assuming that any escape will represent a sequence of the octets in a UTF-8 string.

The comment about email local parts is definitely not right; I'll try to rewrite it in the next few days, but you may not like the results.  In particular, the only "scheme" that, AFAIK, is relevant for email addresses is "mailto:".   It is defined in RFC 6068, which predates the standards track SMTPUTF8 specs and uses the local-part and domain definitions from RFC 5322.  Consequently, as far as "mailto:" is concerned, there is no such thing as non-ASCII characters in the local-part and any IDNs in the domain part have to be (given other specs) transcoded to ASCII using Punycode.  Maybe downhill from there: see the fairly extended rant at 
<https://mailarchive.ietf.org/arch/msg/regext/fiYY81Y7ldmLuWnsHtiZFumx-8M>, supplemented by more ranting at <https://mailarchive.ietf.org/arch/msg/regext/O0TeWi-FP1rouwLtgA2Cqhbtk00>.  There are responses from Martin and Asmus later in the thread that disagree with me to which I have not yet responded.

klensin commented:
Sorry.. I did not read far enough.  RFC 6068 does provide for percent-encoding domain names (which the IDNA specs consider bad practice) and local-parts.  The latter gets UTF-8 and then %-encoding right.  But its last sentence of Section 2, item 5, says 
"When a <local-part>  containing non-ASCII characters will be used to compose a  message, the <local-part> MUST be transformed to conform to  whatever encoding may be defined in a future specification for  the internationalization of email addresses."   That "future specification" is the SMTPUTF8 documents and they, after extensive discussion in the WG, quite explicitly prohibit trying to encode local parts.    The problem goes back to the email specs of the early 1980s and, IIR, somewhat earlier, when the local-part of an email address was specified as allowing absolutely any "printable" ASCII character although some specials and spaces had to be escaped.  RFC 6068 correctly describes that situation but, once one moves past ASCII in the local part, not only is user%host@example.com a valid email address, but so is user%%whatever@example.com and user%25whatever@example.com: there is no way from looking at the things to know whether  they are  actual addresses or represent an encoded form that must be decoded (in case it isn't clear, user%%25whatever@example.com is a valid email address (not encoded) too.

aphillips committed:
Addressing some of @kleinsin's comments

aphillips committed:
Fix one typo

aphillips commented:
Thanks @klensin. I've tried to clean up the top part of the note to be more specific about the relationship of URI `%HH` to character encodings. See what you think.

My goal with the "left hand side" comment was not to claim any defects in the SMTPUTF8 work, but rather to call out that, in spite of standards work being done, interoperability and availability of non-ASCII in email remains incomplete. I have in mind the `input type=email` conversation with WHATWG among other things. Maybe a different example or no example at all would be better?

aphillips committed:
Add percent encoding definition

Replace %HH with the definition and link appropriately.

aphillips committed:
Remove note and mention of email

Having discussed privately with @kleinsin and considered this more deeply, I removed more of the editorial discussion and focused on the current state.

aphillips committed:
Fixing the example markup

As described. Also added the code points for mojibake.

xfq commented on index.html:
```diff
@@ -2748,25 +2748,67 @@ <h3>Basics</h3>
 </ul>
 </aside>
 
+    <p>The situation with regards to specifying support of non-ASCII characters in <a>resource identifiers</a> is complicated because there are at least three specifications ( URI [[RFC3986]], IRI [[RFC3987]], and [[URL]]) that define resource identifiers and their serialization. The WhatWG [[URL]] specification is an attempt to address this complexity by documenting the actual practice of browse ...
```
```suggestion
    <p>The situation with regards to specifying support of non-ASCII characters in <a>resource identifiers</a> is complicated because there are at least three specifications (URI [[RFC3986]], IRI [[RFC3987]], and [[URL]]) that define resource identifiers and their serialization. The WhatWG [[URL]] specification is an attempt to address this complexity by documenting the actual practice of browsers and other user agents. The stated goal of the URL specification is to obsolete both RFCs.</p>
```

Editorial nit.

aphillips committed:
Update index.html

Co-authored-by: Fuqiao Xue <xfq@w3.org>

aphillips committed:
Changes requested in telecon of 2022-10-06

Remove subsection. Move blue box to appropriate location. Update SHOULD->MUST for UTF-8 under 3986 MAY.

aphillips commented:
Merging per teleconference of 2022-10-06 and address [ACTION-1212](https://www.w3.org/International/track/actions/1212).
